hibernate-core/src/main/java/org/hibernate/type/descriptor/java/LocaleTypeDescriptor.java

```diff
@@ -51,43 +51,33 @@ public Locale fromString(String string) {
 			return Locale.ROOT;
 		}
 
-		boolean separatorFound = false;
-		int position = 0;
-		char[] chars = string.toCharArray();
-
-		for ( int i = 0; i < chars.length; i++ ) {
-			// We just look for separators
-			if ( chars[i] == '_' ) {
-				if ( !separatorFound ) {
-					// On the first separator we know that we have at least a language
-					string = new String( chars, position, i - position );
-					position = i + 1;
-				}
-				else {
-					// On the second separator we have to check whether there are more chars available for variant
-					if ( chars.length > i + 1 ) {
-						// There is a variant so add it to the constructor
-						return new Locale( string, new String( chars, position, i - position ), new String( chars,
-								i + 1, chars.length - i - 1 ) );
+		Locale.Builder builder = new Locale.Builder();
+		String[] parts = string.split("_");
```

Comment:
From performance perspective, we should avoid using `String.split()` for it will create RE and might involve unnecessary performance cost. There should be better way to replace the above `split()` usage.

Reply:
Maybe you can use the `StringHelper#split` for better performance.

Reply:
I had originally used [StringTokenizer](https://docs.oracle.com/en/java/javase/14/docs/api/java.base/java/util/StringTokenizer.html), which should have better performance, but its Javadoc says the following:

> StringTokenizer is a legacy class that is retained for compatibility reasons although its use is discouraged in new code. It is recommended that anyone seeking this functionality use the split method of String or the java.util.regex package instead.

That said, even in the latest JDK the class is technically not deprecated, not if the above advice should be heeded.

Looking at the suggested ```StringHelper#split``` method, it actually does use StringTokenizer internally. Depending on how careful we want to be about avoiding extra allocations, we can use StringTokenizer directly and avoid the array allocation. I don't have a strong opinion.

Reply:
Thanks for knowledge sharing. I have no idea that `StringTokenizer` is not recommended. I took a look of the internal implementation of `String#split()` and found it has a fastpath for simple delimiter, so I think your current approach is perfectly fine.
![Screen Shot 2020-06-04 at 4 42 41 PM](https://user-images.githubusercontent.com/8211458/83808229-74936f00-a682-11ea-9827-768121632d2d.png)

Reply:
The `split` method accepts a regex which gets compiled at runtime, it would be best to declare the regex as a static final constant in the class, and use that to split strings.

Reply:
@Sanne as mentioned in the above comment, the split method has a fast path when splitting on a single character that avoids compiling a regex.